drivers/docker/handle.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	// TODO When we expose signals we will need a mapping layer that converts
+	// MacOS signals to the correct signal number for docker. Or we change the
+	// interface to take a signal string and leave it up to driver to map?
+
+	dockerSignal := docker.Signal(sysSig)
+	opts := docker.KillContainerOptions{
+		ID:     h.containerID,
+		Signal: dockerSignal,
+	}
+	return h.client.KillContainer(opts)
+
+}
+
+// Kill is used to terminate the task.
+func (h *taskHandle) Kill(killTimeout time.Duration, signal os.Signal) error {
+	// Only send signal if killTimeout is set, otherwise stop container
+	if killTimeout > 0 {
+		if err := h.Signal(signal); err != nil {
+			return err
+		}
+		select {
+		case <-h.waitCh:
+			return nil
+		case <-time.After(killTimeout):
+		}
+	}
+
+	// Stop the container
+	err := h.client.StopContainer(h.containerID, 0)
+	if err != nil {
+
+		// Container has already been removed.
+		if strings.Contains(err.Error(), NoSuchContainerError) {
+			h.logger.Debug("attempted to stop nonexistent container")
+			return nil
+		}
+		h.logger.Error("failed to stop container", "error", err)
+		return fmt.Errorf("Failed to stop container %s: %s", h.containerID, err)
+	}
+	h.logger.Info("stopped container")
+	return nil
+}
+
+func (h *taskHandle) Stats() (*structs.TaskResourceUsage, error) {
+	h.resourceUsageLock.RLock()
+	defer h.resourceUsageLock.RUnlock()
+	var err error
+	if h.resourceUsage == nil {
+		err = fmt.Errorf("stats collection hasn't started yet")
+	}
+	return h.resourceUsage, err
+}
+
+func (h *taskHandle) run() {
+	exitCode, werr := h.waitClient.WaitContainer(h.containerID)
+	if werr != nil {
+		h.logger.Error("failed to wait for container; already terminated")
+	}
+
+	if exitCode != 0 {
+		werr = fmt.Errorf("Docker container exited with non-zero exit code: %d", exitCode)
+	}
+
+	container, ierr := h.waitClient.InspectContainer(h.containerID)
+	oom := false
+	if ierr != nil {
+		h.logger.Error("failed to inspect container", "error", ierr)
+	} else if container.State.OOMKilled {
+		oom = true
+		werr = fmt.Errorf("OOM Killed")
+	}
+
+	close(h.doneCh)
+
+	// Shutdown the syslog collector
+
+	// Stop the container just incase the docker daemon's wait returned
+	// incorrectly
+	if err := h.client.StopContainer(h.containerID, 0); err != nil {
+		_, noSuchContainer := err.(*docker.NoSuchContainer)
+		_, containerNotRunning := err.(*docker.ContainerNotRunning)
+		if !containerNotRunning && !noSuchContainer {
+			h.logger.Error("error stopping container", "error", err)
+		}
+	}
+
+	// Remove the container
+	if h.removeContainerOnExit == true {
+		if err := h.client.RemoveContainer(docker.RemoveContainerOptions{ID: h.containerID, RemoveVolumes: true, Force: true}); err != nil {
+			h.logger.Error("error removing container", "error", err)
+		}
+	} else {
+		h.logger.Debug("not removing container due to config")
+	}
+
+	// Set the result
+	h.exitResult = &drivers.ExitResult{
```

Comment:
As stated above, I think any fields that are accessed from the `run()` goroutine and other APIs needs guarding with a lock.